File: leetcode/lc-20.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            if c in '([{':
                stack.append(c)
            else:
                if len(stack) == 0:
                    return False
                tmp = stack.pop()
                if c == ')' and tmp != '(':
                    return False
                elif c == ']' and tmp != '[':
                    return False
                elif c == '}' and tmp != '{':
                    return False
        return len(stack) == 0

File: leetcode/lc-20.js
var isValid = function (s) {
  const map = {
    ")": "(",
    "]": "[",
    "}": "{",
  };
  const stack = [];
  for (const c of s) {
    if (c === "(" || c === "[" || c === "{") {
      stack.push(c);
    } else if (stack.length === 0) {
      return false;
    } else if (map[c] !== stack.pop()) {
      return false;
    }
  }
  return stack.length === 0;
};
